### DISCUSSION TITLE
Private Github Rubygems registry for bundler

### DISCUSSION THREAD

**Discussion by allanmc** [renovatebot/renovate#11407] [Request Help]
### How are you running Renovate?

WhiteSource Renovate hosted app on github.com

### Describe the bug

I'm trying to get Renovate to update gems from a private rubygems repository hosted on http://rubygems.pkg.github.com/.
But no matter how i try to configure Renovate, it can't seem to find the updated dependencies.
Is Github hosted Rubygems not supported? 

Minimal reproduction: https://github.com/allanmc/renovate-bundler-issue

### Relevant debug logs

<details><summary>Logs</summary>

```
DEBUG: RubyGems lookup for dependency (sd-rails-utils)(dependency="sd-rails-utils")
DEBUG: Datasource 404
{
  "datasource": "rubygems",
  "lookupName": "sd-rails-utils",
  "url": "https://rubygems.pkg.github.com/mercell/api/v1/gems/sd-rails-utils.json"
}
DEBUG: Failed to look up dependency sd-rails-utils (sd-rails-utils)(packageFile="subfolder/Gemfile", dependency="sd-rails-utils")
```

</details>

### Have you created a minimal reproduction repository?

I have linked to a minimal reproduction repository in the bug description

**Comment by rarkins** [comment]
The question is two parts:
- Is the URL we are calling correct?
- Are the credentials correct?

Based on the brief docs I can find it seems like a Bearer token is required, so your hostRules entry looks correct. Can you reach the url `https://rubygems.pkg.github.com/mercell/api/v1/gems/sd-rails-utils.json` using curl/postman/etc using the same auth?

**Comment by allanmc** [comment]
@rarkins 
The `token` provided is exactly the same as we use for our private npm repository at `npm.pkg.github.com` - the token works fine for this. Should the token for rubygems/bundler be any different? E.g. additional permissions, etc?

I can not seem to reach the URL `https://rubygems.pkg.github.com/mercell/api/v1/gems/sd-rails-utils.json` with the same token - I also get 404.
I'm not sure if Github's Rubygems registry completely follows the Rubygems API?

When running `bundler install`, I can see that bundler fetches the following url:
```
Query Gemcutter Dependency Endpoint API: sd-rails-utils
HTTP GET https://rubygems.pkg.github.com/mercell/api/v1/dependencies?gems=sd-rails-utils
HTTP 200 OK https://rubygems.pkg.github.com/mercell/api/v1/dependencies?gems=sd-rails-utils
```

Fetching that URL works fine with the provided token.

**Comment by allanmc** [comment]
If it's any help, here's the package:
![image](https://user-images.githubusercontent.com/127648/130592982-23c5aea4-ee73-4ab8-8913-b9d92c8f01cb.png)

**Comment by viceice** [comment]
i think this is related to #11107

**Comment by viceice** [comment]
https://github.com/renovatebot/renovate/blob/48acb427b76d69dab21dd5b04674b1ffb50b2d57/lib/datasource/rubygems/releases.ts#L14-L16

https://github.com/renovatebot/renovate/blob/48acb427b76d69dab21dd5b04674b1ffb50b2d57/lib/datasource/rubygems/common.ts#L10

https://github.com/renovatebot/renovate/blob/48acb427b76d69dab21dd5b04674b1ffb50b2d57/lib/datasource/rubygems/get.ts#L13-L41

**Comment by rarkins** [comment]
OK, I will try to test the app on v26 today or tomorrow.

**Comment by viceice** [comment]
@allanmc Can you check the logs which renovate version was used? Should be newer than v26.3.0 to support rubygems on github

your `hostType` is wrong, should be `rubygems`. @rarkins Maybe we need some migrations for `hostType=bundler`  to `hostType=rubygems` ?

https://github.com/renovatebot/renovate/blob/48acb427b76d69dab21dd5b04674b1ffb50b2d57/lib/datasource/rubygems/common.ts#L7-L8

**Comment by allanmc** [comment]
> @allanmc Can you check the logs which renovate version was used? Should be newer than v26.3.0 to support rubygems on github

It's running:
```
INFO: Repository started
{
  "renovateVersion": "25.76.2"
}
```
> your hostType is wrong, should be rubygems. 

I can switch to `hostType=rubygems`, but i just tried to follow this part of the documentation:
https://docs.renovatebot.com/modules/manager/bundler/
It says:
> hostType=bundler is essential. If the registry is used for multiple package types then you may need multiple hostRules. You cannot leave off hostType.

**Comment by viceice** [comment]
@rarkins I think we can move this to a discussion, as it's not a bug. The app and the docs supports this, but the app isn't updated to the required version yet.

**Comment by rarkins** [comment]
We had incorrect documentation for bundler hostType previously.

**Comment by viceice** [comment]
yes, and it's now wrong again 🙈 @HonkingGoose Can you suggest a docs change please? 

https://github.com/renovatebot/renovate/blob/f00413c298ad63e1ff0bf03d4dba7c1db3ca5be4/lib/manager/bundler/readme.md

`hostType` needs to be `rubygems` on v26 / current main

**Reply by HonkingGoose** [reply]
See PR:

- https://github.com/renovatebot/renovate/pull/11413

**Comment by thomasLeclaire** [comment]
Hello! I have an issue which sounds similar. using the well-known sidekiq gems, there is a private repo for the pro&enterprise versions. 
but https://enterprise.contribsys.com/api/v1/dependencies?gems=.. doesn't exist, resulting into some 404 error on renovate side. 

but the "old" versions of rubygems endpoints are available : https://github.com/mperham/sidekiq/issues/4988
https://enterprise.contribsys.com/info/sidekiq-ent

could be good that renovate auto fallback to this old but still used configurations.

**Reply by viceice** [reply]
@thomasLeclaire `/info/<gemName>` isn't a known bundler api. We currently support these api's for selfhosted registries:

https://github.com/renovatebot/renovate/blob/48acb427b76d69dab21dd5b04674b1ffb50b2d57/lib/datasource/rubygems/get.ts#L11-L13

**Reply by viceice** [reply]
I think we need some api specs first, then we can add that api to our rubygems datasource

**Reply by thomasLeclaire** [reply]
Hello, 
below the answers of the maintainers of this system : 

Conceptually, probably this blog post: https://andre.arko.net/2014/03/28/the-new-rubygems-index-format/

Technically, this repo, where we keep the reference implementation that RubyGems.org uses: https://github.com/rubygems/compact_index

**Reply by rarkins** [reply]
I think those URLs were more about helping the sidekiq maintainer to *add* API compatibility to his server. In our case, it sounds like we should see if we can loosen our requirements on the `gems` and `dependencies` endpoints if those don't exist

**Reply by narwold** [reply]
The issue that @thomasLeclaire opened up on the sidekiq repo is here for reference: https://github.com/mperham/sidekiq/issues/4988

This is currently a blocker for my company using Renovate on the Ruby side.

**Reply by rarkins** [reply]
OK, can someone create a feature request (assuming one doesn't exist) for us to support the "old" index as identified in https://github.com/renovatebot/renovate/discussions/11407#discussioncomment-1300392

I think it looks like what we have hardcoded to use for rubygems.org, so it's hopefully not much work.

Reproduction essential (e.g. using sidekiq)

**Reply by narwold** [reply]
Sure

**Reply by narwold** [reply]
https://github.com/renovatebot/renovate/issues/16126
